pyscf/adc/uadc.py

```diff
@@ -49,33 +49,32 @@ def kernel(adc, nroots=1, guess=None, eris=None, verbose=None):
 
     guess = adc.get_init_guess(nroots, diag, ascending = True)
 
-    conv, E, U = lib.linalg_helper.davidson1(lambda xs : [matvec(x) for x in xs], guess, diag, nroots=nroots, verbose=log, tol=adc.conv_tol, max_cycle=adc.max_cycle, max_space=adc.max_space)
+    conv, adc.E, adc.U = lib.linalg_helper.davidson1(lambda xs : [matvec(x) for x in xs], guess, diag, nroots=nroots, verbose=log, tol=adc.conv_tol, max_cycle=adc.max_cycle, max_space=adc.max_space,tol_residual=adc.tol_residual)
 
-    U = np.array(U)
+    if adc.compute_properties == True:
+        adc.P,adc.X_a,adc.X_b = adc.get_properties(nroots)
 
-    T = adc.get_trans_moments()
+    nfalse = np.shape(conv)[0] - np.sum(conv)
 
-    spec_factors = adc.get_spec_factors(T, U, nroots)
+    logger.info(adc, "\n*************************************************************")
+    logger.info(adc, "                  ADC calculation summary")
+    logger.info(adc, "*************************************************************")
 
-    nfalse = np.shape(conv)[0] - np.sum(conv)
     if nfalse >= 1:
-        print ("*************************************************************")
-        print (" WARNING : ", "Davidson iterations for ",nfalse, "root(s) not converged")
-        print ("*************************************************************")
+        logger.info(adc, "*************************************************************")
+        logger.info(adc, " WARNING : ", "Davidson iterations for ", nfalse, "root(s) not converged")
+        logger.info(adc, "*************************************************************")
 
-
-    if adc.verbose >= logger.INFO:
-        if nroots == 1:
-            logger.info(adc, '%s root %d    Energy (Eh) = %.10f    Energy (eV) = %.8f    Spec factors = %.8f    conv = %s',
-                         adc.method, 0, E, E*27.2114, spec_factors, conv)
-        else :
-            for n, en, pn, convn in zip(range(nroots), E, spec_factors, conv):
-                logger.info(adc, '%s root %d    Energy (Eh) = %.10f    Energy (eV) = %.8f    Spec factors = %.8f    conv = %s',
-                          adc.method, n, en, en*27.2114, pn, convn)
+    for n in range(nroots):
+        print_string = ('%s root %d  |  Energy (Eh) = %10.10f  |  Energy (eV) = %10.8f  ' % (adc.method, n, adc.E[n], adc.E[n]*27.2114))
+        if adc.compute_properties == True:
+            print_string += ("|  Spec factors = %10.8f  " % adc.P[n])
+        print_string += ("|  conv = %s" % conv[n])
+        logger.info(adc, print_string)
 
     log.timer('ADC', *cput0)
 
-    return E, U, spec_factors
+    return adc.E, adc.U, adc.P, adc.X_a, adc.X_b
```

Comment:
U is already a tuple of two arrays. We should do the same for all other arrays, including X. Please make sure that you are return a tuple of X_a and X_b (needs to be defined prior) and adjust the rest of the code, including tests and examples.

Reply:
Done